packages/automerge-doc-server/src/socket.ts

```diff
@@ -1,37 +1,60 @@
 import { type Socket, io } from "socket.io-client";
 
 import type { JsonValue, RefContent } from "../../backend/pkg/src/index.ts";
+import type { NewDocSocketResponse, StartListeningSocketResponse } from "./types.js";
 
 /** Messages handled by the `SocketServer`. */
-export type Handlers = {
-    create_doc: (data: RefContent, callback: (docId: string) => void) => void;
-    get_doc: (refId: string, callback: (docId: string | null) => void) => void;
+export type SocketIOHandlers = {
+    createDoc: (data: RefContent) => Promise<NewDocSocketResponse>;
+    cloneDoc: (docId: string) => Promise<NewDocSocketResponse>;
+    startListening: (refId: string, docId: string) => Promise<StartListeningSocketResponse>;
 };
 
 /** Messages emitted by the `SocketServer`. */
 export type Requests = {
     autosave: (data: RefContent) => void;
 };
 
-/** Encapsulates socket.io for internal communication with the backend.
+// Convert the type of each property from `(...args) => ret` to `(...args, (ret) => void) => void` to
+// match the signature expected for SocketIO listeners
+type Callbackify<T> = {
+    [K in keyof T]: T[K] extends (...args: infer Args) => infer Ret
+        ? (...args: [...Args, (res: Promise<Ret>) => void]) => void
+        : never;
+};
+
+function registerHandler<
```

Comment:
I got annoyed at all the plumbing I had to do when modifying the handlers so I removed all the biolerplate-like code.